docs/src/abigen/the-abigen-macro.md

```diff
@@ -0,0 +1,106 @@
+# abigen!
+
+`abigen!` is a procedural macro -- it generates code. It accepts inputs in the format of:
+```text
+ProgramType(name="MyProgramType", abi="my_program-abi.json")...
+```
+where:
+
+`ProgramType` is one of: `Contract`, `Script` or `Predicate`,
+
+`name` is the name that will be given to the generated bindings,
+
+`abi` is either a path to the json abi file or its actual contents.
+
+---
+So, an `abigen!` which generates bindings for two contracts and one script looks like this:
+```rust,ignore
+{{#include ../../../examples/abigen/src/lib.rs:multiple_abigen_program_types}}
+```
+
+## How does the generated code look?
+
+A rough overview:
+```rust,ignore
+pub mod abigen_bindings {
+    pub mod contract_a_mod {
+        struct SomeCustomStruct{/*...*/};
+        // other custom types used in the contract
+        
+        struct ContractA {/*...*/};
+        impl ContractA {/*...*/};
+        // ...
+    }
+    pub mod contract_b_mod {
+        // ...
+    }
+    pub mod my_script_mod {
+        // ...
+    }
+    pub mod my_predicate_mod{
+        // ...
+    }
+    pub mod shared_types{
+        // ...
+    }
+}
+
+pub use contract_a_mod::{/*..*/};
+pub use contract_b_mod::{/*..*/};
+pub use my_predicate_mod::{/*..*/};
+pub use shared_types::{/*..*/};
+```
+
+Each `ProgramType` gets its own `mod` based on the `name` given in the `abigen!`. Inside the respective mods the custom
+types used by that program are generated, as well as the bindings through which the actual calls can be made.
+
+There is one extra `mod` called `shared_types` which is generated if `abigen!` detects that the given programs share
+types. Instead of each `mod` regenerating the type for itself, the type is lifted out into the `shared_types` module and
+generated only once and then shared between all programs that use it.
+
+A type is deemed shared if its name and definition match up. This can happen either because you've used the same library
+(a custom one or a type from the stdlib) or because you've happened to define the exact same type.
+
+Finally `pub use` statements are inserted, so you don't have to fully qualify the generated types. To avoid conflict
+only types which have unique names will get a `pub use` statement. If you find rustc can't find your type it might just
+be that there is another generated type with the same name. To fix the issue just qualify the path by doing
+`abigen_bindings::whatever_contract_mod::TheType`.
```

Comment:
Same here: no break lines in a paragraph + wording improvements.

```suggestion
Finally, `pub use` statements are inserted, so you don't have to fully qualify the generated types. To avoid conflict, only types that have unique names will get a `pub use` statement. If you find rustc can't find your type, it might just be that there is another generated type with the same name. To fix the issue just qualify the path by doing `abigen_bindings::whatever_contract_mod::TheType`.
```